chain/src/chain_service.rs

```diff
@@ -0,0 +1,156 @@
+
+/// Chain background service to receive LonelyBlock and only do `non_contextual_verify`
+pub(crate) struct ChainService {
+    shared: Shared,
+
+    process_block_rx: Receiver<ProcessBlockRequest>,
+
+    orphan_broker: OrphanBroker,
+}
+impl ChainService {
+    /// Create a new ChainService instance with shared.
+    pub(crate) fn new(
+        shared: Shared,
+        process_block_rx: Receiver<ProcessBlockRequest>,
+
+        consume_orphan: OrphanBroker,
+    ) -> ChainService {
+        ChainService {
+            shared,
+            process_block_rx,
+            orphan_broker: consume_orphan,
+        }
+    }
+
+    /// Receive block from `process_block_rx` and do `non_contextual_verify`
+    pub(crate) fn start_process_block(self) {
+        let signal_receiver = new_crossbeam_exit_rx();
+
+        let clean_expired_orphan_timer =
+            crossbeam::channel::tick(std::time::Duration::from_secs(60));
+
+        loop {
+            select! {
+                recv(self.process_block_rx) -> msg => match msg {
+                    Ok(Request { responder, arguments: lonely_block }) => {
+                        // asynchronous_process_block doesn't interact with tx-pool,
+                        // no need to pause tx-pool's chunk_process here.
+                        let _trace_now = minstant::Instant::now();
+                        self.asynchronous_process_block(lonely_block);
+                        if let Some(handle) = ckb_metrics::handle(){
+                            handle.ckb_chain_async_process_block_duration.observe(_trace_now.elapsed().as_secs_f64())
+                        }
+                        let _ = responder.send(());
+                    },
+                    _ => {
+                        error!("process_block_receiver closed");
+                        break;
+                    },
+                },
+                recv(clean_expired_orphan_timer) -> _ => {
+                    self.orphan_broker.clean_expired_orphans();
+                },
+                recv(signal_receiver) -> _ => {
+                    info!("ChainService received exit signal, exit now");
+                    break;
+                }
+            }
+        }
+    }
+
+    fn non_contextual_verify(&self, block: &BlockView) -> Result<(), Error> {
+        let consensus = self.shared.consensus();
+        BlockVerifier::new(consensus).verify(block).map_err(|e| {
+            debug!("[process_block] BlockVerifier error {:?}", e);
+            e
+        })?;
+
+        NonContextualBlockTxsVerifier::new(consensus)
+            .verify(block)
+            .map_err(|e| {
+                debug!(
+                    "[process_block] NonContextualBlockTxsVerifier error {:?}",
+                    e
+                );
+                e
+            })
+            .map(|_| ())
+    }
+
+    // `self.non_contextual_verify` is very fast.
+    fn asynchronous_process_block(&self, lonely_block: LonelyBlock) {
```

Comment:
seems `asynchronous_` is not the best name for this function, which makes me think of async fn in tokio :)
it's more like `preprocess_block`, we only do `non_contextual_verify` and insert into db.